samples/image-uploader/src/utils.js

```diff
@@ -0,0 +1,89 @@
+import imageMimeTypes from "./image-mime-types.json"
+
+/* Takes an HTML5 file, encodes as Base64 data Url,
+   returns a two element list of Base64 Url prefix and
+   raw Base64 data.
+
+   readFileAsUrl(file: HTMLFile): [string, string]
+ */
+export function readFileAsUrl(file) {
+  return new Promise((resolve, reject) => {
+    const reader = new FileReader()
+
+    reader.onload = function(e) {
+      resolve(e.target.result.split(","))
+    }
+
+    reader.onerror = function(e) {
+      reject(e.error)
+    }
+
+    reader.onabort = function(e) {
+      reject(new Error("File aborted."))
+    }
+
+    reader.readAsDataURL(file)
+  })
+}
+
+// getImageUrlFromDataTransfer(dataTransfer: DataTransfer): string?
+export function getImageUrlFromDataTransfer(dataTransfer) {
+  const html = dataTransfer.getData("text/html")
+  if (!html) {
+    return
+  }
+
+  const frag = document.createElement("main")
+  frag.innerHTML = html
+
+  return frag.querySelector("img").src
+}
+
+// getAssetIdFromDataTransfer(dataTransfer: DataTransfer): string?
+export function getAssetIdFromDataTransfer(dataTransfer) {
+  const plain = dataTransfer.getData("text/plain")
+  const match = plain.match(/spaces\/\w+\/assets\/(\w+)\/?$/)
+
+  if (!match) {
+    return
+  }
+
+  return match[1]
+}
+
+// getBase64FromDataTransfer(dataTransfer: DataTransfer): [base64Prefix, base64Data]
+export function getBase64FromDataTransfer(dataTransfer) {
+  const src = getImageUrlFromDataTransfer(dataTransfer)
+  if (!/^data:image\/\w+;base64,/.test(src)) {
+    return
+  }
+
+  const parts = src.split(",")
+
+  return {
+    prefix: parts[0],
+    data: parts[1],
+    type: src.split(/:|;/)[1]
+  }
+}
+
+// getMimeTypeByPath(path: string): [mimetype, isImage]
+export function getMimeTypeByPath(uri) {
+  // Unsplash doesn't have file extensions in their urls, so
+  if (isUnsplashImageUrl(uri)) {
```

Comment:
isn't it a too specific check? there are thousands of websites

Reply:
How about we fire a `HEAD` request and look at the `Content-Type` header if we cannot infer the mime type from the url?

Reply:
Removed mime type sniffing in favor of relying on `DataTransfer` content